utils/src/storage.rs

```diff
@@ -0,0 +1,137 @@
+use std::fs::{DirBuilder, File, ReadDir};
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+
+use bincode;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub struct FsStorageHelper {
+    root: PathBuf,
+}
+
+impl FsStorageHelper {
+    pub fn new(root_path: &Path) -> Result<Self, std::io::Error> {
+        DirBuilder::new()
+            .recursive(true)
+            .create(root_path)
+            .map(|_| Self {
+                root: root_path.to_path_buf(),
+            })
+    }
+
+    fn get_root_path(&self, path: Option<&Path>) -> PathBuf {
+        match path {
+            Some(p) => p.to_path_buf(),
+            None => self.root.clone(),
+        }
+    }
+
+    pub fn add_folder(
+        &self,
+        folder_name: &str,
+        path: Option<&Path>,
+    ) -> Result<PathBuf, std::io::Error> {
+        let path = self.get_root_path(path);
+        let folder_path = path.join(folder_name);
+        DirBuilder::new()
+            .recursive(true)
+            .create(folder_path.clone())
+            .map(|_| folder_path)
+    }
+
+    pub fn iterate_folder(
+        &self,
+        folder_name: &str,
+        path: Option<&Path>,
+    ) -> Result<ReadDir, std::io::Error> {
+        let path = self.get_root_path(path);
+        let folder_path = path.join(folder_name);
+        folder_path.read_dir()
+    }
+
+    pub fn search_item(
+        &self,
+        path: Option<&Path>,
+        item_name: &str,
+        extension: Option<&str>,
+    ) -> Result<Option<PathBuf>, std::io::Error> {
+        let path = self.get_root_path(path);
+        if path.is_dir() {
+            return Ok(path
+                .read_dir()?
+                .flat_map(|res| res.map(|e| e.path()))
+                .find(|el| {
+                    let ok_name = el
+                        .file_stem()
+                        .and_then(|os| os.to_str())
+                        .filter(|&name| name == item_name)
+                        .is_some();
+                    let ok_ext = match extension {
+                        Some(ext) => el
+                            .extension()
+                            .and_then(|os| os.to_str())
+                            .filter(|&extension| extension == ext)
+                            .is_some(),
+                        None => true,
+                    };
+                    ok_name && ok_ext
+                }));
+        }
```

Comment:
```rust
} else {
    Ok(None)
}
```

Reply:
I prefer not to put the else because there is only one way this function should work: with directories.
Putting else means that there is some other work to do, but here there is none. Just style preference I guess.

Reply:
Then shouldn't you return an error

Reply:
or if you don't care about returning errors for the other 2 errors read_dir can produce, you can just match on read_dir, I don't really see a problem here anyway